mesos-rxjava-core/src/main/java/org/apache/mesos/rx/java/MesosSchedulerClient.java

```diff
@@ -93,7 +116,43 @@ public MesosSchedulerClient(
         return createPost.call(subscription)
             .flatMap(httpClient::submit)
             .subscribeOn(Schedulers.io())
-            .flatMap(AbstractHttpContentHolder::getContent)
+            .flatMap(resp -> {
+                final HttpResponseStatus status = resp.getStatus();
+                final int code = status.code();
+
+                if (code == 200) {
+                    return resp.getContent();
+                } else {
+                    //TODO: Figure out how to get these subscribe exceptions to propagate up so that the observable dies
+                    final HttpResponseHeaders headers = resp.getHeaders();
+                    final List<Map.Entry<String, String>> entries = headers.entries();
+                    String errorMessage;
+                    if (headers.getContentLength() > 0) {
+                        errorMessage = resp.getContent()
+                            .map(r -> r.toString(CharsetUtil.UTF_8))
+                            .toBlocking()
+                            .first();
+                    } else {
+                        errorMessage = "";
+                    }
+
+                    final MesosClientErrorContext context = new MesosClientErrorContext(code, errorMessage, entries);
+                    if (400 <= code && code < 500) {
+                        // client error
+                        throw new MesosClientException(subscription, context);
+                    } else if (500 <= code && code < 600) {
+                        // server error
+                        throw new MesosServerException(subscription, context);
+                    } else {
+                        // Unknown error
+                        LOGGER.warn("Unhandled error: context = {}", context);
+                        // This shouldn't actually ever happen, but it's here for completeness of the if-else tree
```

Comment:
yeah, we'll need to clean this part up in a new PR. this code has now been duplicated here and in SinkSubscriber.